File: src/render.zig
//! Screen rendering
//! Full-screen redraw with viewport management

const std = @import("std");
const buffer = @import("buffer.zig");
const backend = @import("term/backend.zig");
const mode = @import("mode.zig");
const key = @import("key.zig");

/// Viewport tracks visible lines
pub const Viewport = struct {
    top_line: usize,
    visible_rows: u16,
    visible_cols: u16,

    pub fn init(visible_rows: u16, visible_cols: u16) Viewport {
        return Viewport{
            .top_line = 0,
            .visible_rows = visible_rows,
            .visible_cols = visible_cols,
        };
    }

    /// Update viewport to ensure cursor line is visible
    pub fn update(self: *Viewport, cursor_line: usize, total_lines: usize) void {
        // If cursor is above viewport, scroll up
        if (cursor_line < self.top_line) {
            self.top_line = cursor_line;
        }

        // If cursor is below viewport, scroll down
        const bottom_line = self.top_line + self.visible_rows - 1;
        if (cursor_line > bottom_line) {
            self.top_line = cursor_line - (self.visible_rows - 1);
            if (self.top_line > total_lines) {
                self.top_line = if (total_lines > 0) total_lines - 1 else 0;
            }
        }

        // Ensure top_line doesn't go negative
        if (self.top_line > total_lines) {
            self.top_line = if (total_lines > 0) total_lines - 1 else 0;
        }
    }
};

/// Render the entire screen
pub fn render(
    term: *backend.Terminal,
    buf: *buffer.GapBuffer,
    viewport: *Viewport,
    allocator: std.mem.Allocator,
    current_mode: mode.Mode,
    pending_key: ?key.Key,
) !void {
    // Clear screen
    try term.clearScreen();

    // Update viewport
    const cursor_line = buf.getLine();
    const total_lines = buf.getTotalLines();
    viewport.update(cursor_line, total_lines);

    const before_gap = buf.getBeforeGap();
    const after_gap = buf.getAfterGap();

    // Render visible lines by iterating through content
    // Reserve last row for status bar
    const content_rows = viewport.visible_rows - 1;
    const start_line = viewport.top_line;
    var line_num: usize = 0;
    var display_row: u16 = 0; // Tracks which visible row we're rendering (0 to content_rows-1)
    var line_buffer = std.ArrayList(u8).empty;
    defer line_buffer.deinit(allocator);

    // Iterate through content to find and render lines
    var before_pos: usize = 0;
    var after_pos: usize = 0;
    var in_before_gap = true;

    while (line_num < start_line + content_rows) {
        line_buffer.clearRetainingCapacity();

        // Read one line
        var line_complete = false;
        while (!line_complete) {
            var byte: u8 = undefined;
            var at_end = false;

            if (in_before_gap) {
                if (before_pos < before_gap.len) {
                    byte = before_gap[before_pos];
                    before_pos += 1;
                    if (byte == '\n') {
                        line_complete = true;
                    } else {
                        try line_buffer.append(allocator, byte);
                    }
                } else {
                    in_before_gap = false;
                    if (after_pos < after_gap.len) {
                        byte = after_gap[after_pos];
                        after_pos += 1;
                        if (byte == '\n') {
                            line_complete = true;
                        } else {
                            try line_buffer.append(allocator, byte);
                        }
                    } else {
                        at_end = true;
                        line_complete = true;
                    }
                }
            } else {
                if (after_pos < after_gap.len) {
                    byte = after_gap[after_pos];
                    after_pos += 1;
                    if (byte == '\n') {
                        line_complete = true;
                    } else {
                        try line_buffer.append(allocator, byte);
                    }
                } else {
                    at_end = true;
                    line_complete = true;
                }
            }

            if (at_end and line_buffer.items.len == 0) {
                // End of content
                break;
            }
        }

        // Render this line if it's in viewport
        if (line_num >= start_line) {
            // Check if we've rendered all visible content rows (excluding status bar)
            if (display_row >= content_rows) {
                break; // Past visible area
            }
            
            try term.moveCursor(display_row, 0);

            const line = line_buffer.items;
            const line_to_display = if (line.len > viewport.visible_cols) line[0..viewport.visible_cols] else line;
            try term.write(line_to_display);

            // Clear to end of line
            if (line.len < viewport.visible_cols) {
                try term.write("\x1b[K");
            }

            display_row += 1;
        }

        line_num += 1;

        // Stop if we've rendered all content and filled screen
        if (line_buffer.items.len == 0 and before_pos >= before_gap.len and after_pos >= after_gap.len) {
            break;
        }
    }

    // Clear remaining lines (except status bar)
    while (display_row < content_rows) {
        try term.moveCursor(display_row, 0);
        try term.write("\x1b[K");
        display_row += 1;
    }

    // Render status bar on last line
    try renderStatusBar(term, viewport, current_mode, pending_key);

    // Position cursor (accounting for status bar)
    const cursor_line_in_viewport = if (cursor_line >= viewport.top_line and cursor_line < start_line + content_rows) 
        cursor_line - viewport.top_line 
    else 
        0;
    const cursor_col = calculateCursorColumn(buf, cursor_line);
    const display_col = if (cursor_col < viewport.visible_cols - 1) cursor_col else viewport.visible_cols - 1;
    try term.moveCursor(@as(u16, @intCast(cursor_line_in_viewport)), @as(u16, @intCast(display_col)));
}

/// Render status bar at the bottom of the screen
fn renderStatusBar(
    term: *backend.Terminal,
    viewport: *Viewport,
    current_mode: mode.Mode,
    pending_key: ?key.Key,
) !void {
    const status_row = viewport.visible_rows - 1;
    try term.moveCursor(status_row, 0);
    
    // Invert colors for status bar (reverse video)
    try term.write("\x1b[7m");
    
    // Mode indicator
    const mode_str = switch (current_mode) {
        .normal => "NORMAL",
        .insert => "INSERT",
    };
    try term.write(mode_str);
    
    // Pending key indicator
    if (pending_key) |pending| {
        try term.write(" [");
        const key_str = switch (pending) {
            .char => |ch| blk: {
                var buf: [1]u8 = undefined;
                buf[0] = ch;
                break :blk buf[0..];
            },
            .backspace => "BACKSPACE",
            .enter => "ENTER",
            .escape => "ESC",
            .arrow_up => "UP",
            .arrow_down => "DOWN",
            .arrow_left => "LEFT",
            .arrow_right => "RIGHT",
            .home => "HOME",
            .end => "END",
            .page_up => "PGUP",
            .page_down => "PGDN",
            .delete => "DEL",
            .ctrl => |ch| blk: {
                var buf: [32]u8 = undefined;
                const len = std.fmt.bufPrint(&buf, "CTRL-{}", .{ch}) catch "CTRL-?";
                break :blk len;
            },
        };
        try term.write(key_str);
        try term.write("]");
    }
    
    // Fill rest of line with spaces
    const mode_len = mode_str.len;
    var used_cols = mode_len;
    if (pending_key) |pending| {
        // Calculate length of pending key string
        const key_str_len: usize = switch (pending) {
            .char => 1,
            .backspace => 9,
            .enter => 5,
            .escape => 3,
            .arrow_up => 2,
            .arrow_down => 2,
            .arrow_left => 4,
            .arrow_right => 5,
            .home => 4,
            .end => 3,
            .page_up => 4,
            .page_down => 4,
            .delete => 3,
            .ctrl => 6,
        };
        used_cols += 3 + key_str_len; // "[key]"
    }
    const remaining_cols = if (viewport.visible_cols > used_cols) viewport.visible_cols - used_cols else 0;
    
    var i: usize = 0;
    while (i < remaining_cols) : (i += 1) {
        try term.write(" ");
    }
    
    // Reset colors
    try term.write("\x1b[0m");
}

/// Calculate cursor column position
fn calculateCursorColumn(buf: *buffer.GapBuffer, line: usize) usize {
    const before_gap = buf.getBeforeGap();
    const after_gap = buf.getAfterGap();

    // Count lines and find column
    var current_line: usize = 0;
    var col: usize = 0;
    var pos: usize = 0;

    // Count through before_gap
    for (before_gap) |byte| {
        if (byte == '\n') {
            if (current_line == line) {
                return col;
            }
            current_line += 1;
            col = 0;
        } else {
            col += 1;
        }
        pos += 1;
    }

    // If we're at the gap position
    if (current_line == line) {
        return col;
    }

    // Count through after_gap
    for (after_gap) |byte| {
        if (byte == '\n') {
            if (current_line == line) {
                return col;
            }
            current_line += 1;
            col = 0;
        } else {
            col += 1;
        }
    }

    return col;
}

File: src/render.rs
//! Rendering system
//! Handles drawing the editor UI to the terminal

use crate::buffer::GapBuffer;
use crate::mode::Mode;
use crate::key::Key;
use crate::term::TerminalBackend;
use crate::viewport::Viewport;
use crate::state::State;

/// Render the editor interface
pub fn render<T: TerminalBackend>(
    term: &mut T,
    buf: &GapBuffer,
    viewport: &mut Viewport,
    current_mode: Mode,
    pending_key: Option<Key>,
    state: &State,
) -> Result<(), String> {
    // Clear screen
    term.clear_screen()?;

    // Update viewport based on cursor position
    let cursor_line = buf.get_line();
    let total_lines = buf.get_total_lines();
    viewport.update(cursor_line, total_lines);

    // Render content
    render_content(term, buf, viewport)?;

    // Render status bar
    render_status_bar(term, viewport, current_mode, pending_key, state)?;

    // Position cursor
    let cursor_line_in_viewport = if cursor_line >= viewport.top_line() 
        && cursor_line < viewport.top_line() + viewport.visible_rows() {
        cursor_line - viewport.top_line()
    } else {
        0
    };
    
    let cursor_col = calculate_cursor_column(buf, cursor_line);
    let display_col = cursor_col.min(viewport.visible_cols().saturating_sub(1));
    
    term.move_cursor(cursor_line_in_viewport as u16, display_col as u16)?;
    
    Ok(())
}

fn render_content<T: TerminalBackend>(
    term: &mut T,
    buf: &GapBuffer,
    viewport: &Viewport,
) -> Result<(), String> {
    let before_gap = buf.get_before_gap();
    let after_gap = buf.get_after_gap();
    
    // Combine before and after gap to get full text
    let mut lines: Vec<Vec<u8>> = Vec::new();
    let mut current_line = Vec::new();
    
    // Process before_gap
    for &byte in before_gap {
        if byte == b'\n' {
            lines.push(current_line);
            current_line = Vec::new();
        } else {
            current_line.push(byte);
        }
    }
    
    // Process after_gap
    for &byte in after_gap {
        if byte == b'\n' {
            lines.push(current_line);
            current_line = Vec::new();
        } else {
            current_line.push(byte);
        }
    }
    // Add last line if not empty
    if !current_line.is_empty() || lines.is_empty() {
        lines.push(current_line);
    }
    
    // Render visible lines
    let top_line = viewport.top_line();
    let visible_rows = viewport.visible_rows().saturating_sub(1); // Reserve one row for status bar
    let visible_cols = viewport.visible_cols();
    
    for i in 0..visible_rows {
        let line_num = top_line + i;
        if line_num < lines.len() {
            let line = &lines[line_num];
            // Truncate line to visible width
            let display_line: Vec<u8> = line.iter()
                .take(visible_cols)
                .copied()
                .collect();
            
            // Pad with spaces if line is shorter than visible width
            let mut padded_line = display_line;
            while padded_line.len() < visible_cols {
                padded_line.push(b' ');
            }
            
            term.write(&padded_line)?;
            term.write(b"\r\n")?;
        } else {
            // Empty line
            term.write(&vec![b' '; visible_cols])?;
            term.write(b"\r\n")?;
        }
    }
    
    Ok(())
}

fn render_status_bar<T: TerminalBackend>(
    term: &mut T,
    viewport: &Viewport,
    current_mode: Mode,
    pending_key: Option<Key>,
    state: &State,
) -> Result<(), String> {
    let status_row = viewport.visible_rows().saturating_sub(1);
    term.move_cursor(status_row as u16, 0)?;
    
    // Invert colors for status bar (reverse video)
    term.write(b"\x1b[7m")?;
    
    // Mode indicator
    let mode_str = match current_mode {
        Mode::Normal => "NORMAL",
        Mode::Insert => "INSERT",
    };
    
    term.write(mode_str.as_bytes())?;
    
    // Pending key indicator
    if let Some(key) = pending_key {
        term.write(b" [")?;
        let key_str = format_key(key);
        term.write(key_str.as_bytes())?;
        term.write(b"]")?;
    }
    
    // Debug information (if debug mode is enabled)
    if state.debug_mode {
        let mut debug_parts = Vec::new();
        
        // Last keypress
        if let Some(key) = state.last_keypress {
            debug_parts.push(format!("Last: {}", format_key(key)));
        }
        
        // Cursor position
        debug_parts.push(format!("Pos: {}:{}", state.cursor_pos.0 + 1, state.cursor_pos.1 + 1));
        
        // Buffer stats
        debug_parts.push(format!("Lines: {}", state.total_lines));
        debug_parts.push(format!("Size: {}B", state.buffer_size));
        
        // Join debug parts
        let debug_str = debug_parts.join(" | ");
        
        // Calculate available space
        let mode_len = mode_str.len();
        let pending_len = if pending_key.is_some() { 
            format_key(pending_key.unwrap()).len() + 3 // "[key]"
        } else { 
            0 
        };
        let used_cols = mode_len + pending_len;
        let available_cols = viewport.visible_cols().saturating_sub(used_cols);
        
        // Truncate debug string if needed
        let debug_display = if debug_str.len() <= available_cols {
            debug_str
        } else {
            format!("{}...", &debug_str[..available_cols.saturating_sub(3)])
        };
        
        // Add spacing before debug info
        let spacing = available_cols.saturating_sub(debug_display.len());
        for _ in 0..spacing {
            term.write(b" ")?;
        }
        
        term.write(debug_display.as_bytes())?;
    }
    
    // Fill rest of line with spaces
    let mode_len = mode_str.len();
    let pending_len = if pending_key.is_some() { 
        format_key(pending_key.unwrap()).len() + 3 // "[key]"
    } else { 
        0 
    };
    let debug_len = if state.debug_mode {
        // Calculate actual debug length
        let mut debug_parts = Vec::new();
        if let Some(key) = state.last_keypress {
            debug_parts.push(format!("Last: {}", format_key(key)));
        }
        debug_parts.push(format!("Pos: {}:{}", state.cursor_pos.0 + 1, state.cursor_pos.1 + 1));
        debug_parts.push(format!("Lines: {}", state.total_lines));
        debug_parts.push(format!("Size: {}B", state.buffer_size));
        let debug_str = debug_parts.join(" | ");
        let available_cols = viewport.visible_cols().saturating_sub(mode_len + pending_len);
        debug_str.len().min(available_cols)
    } else {
        0
    };
    let used_cols = mode_len + pending_len + debug_len;
    let remaining_cols = viewport.visible_cols().saturating_sub(used_cols);
    
    for _ in 0..remaining_cols {
        term.write(b" ")?;
    }
    
    // Reset colors
    term.write(b"\x1b[0m")?;
    
    Ok(())
}

fn format_key(key: Key) -> String {
    match key {
        Key::Char(ch) => {
            if ch >= 32 && ch < 127 {
                format!("{}", ch as char)
            } else {
                format!("\\x{:02x}", ch)
            }
        }
        Key::Ctrl(ch) => format!("Ctrl+{}", (ch as char).to_uppercase()),
        Key::ArrowUp => "↑".to_string(),
        Key::ArrowDown => "↓".to_string(),
        Key::ArrowLeft => "←".to_string(),
        Key::ArrowRight => "→".to_string(),
        Key::Backspace => "Backspace".to_string(),
        Key::Delete => "Delete".to_string(),
        Key::Enter => "Enter".to_string(),
        Key::Escape => "Esc".to_string(),
        Key::Tab => "Tab".to_string(),
        Key::Home => "Home".to_string(),
        Key::End => "End".to_string(),
        Key::PageUp => "PageUp".to_string(),
        Key::PageDown => "PageDown".to_string(),
    }
}

fn calculate_cursor_column(buf: &GapBuffer, line: usize) -> usize {
    let before_gap = buf.get_before_gap();
    let mut current_line = 0;
    let mut col = 0;
    
    for &byte in before_gap {
        if byte == b'\n' {
            if current_line == line {
                return col;
            }
            current_line += 1;
            col = 0;
        } else {
            col += 1;
        }
    }
    
    // If we're at the gap position on the target line
    if current_line == line {
        return col;
    }
    
    // Check after_gap
    let after_gap = buf.get_after_gap();
    for &byte in after_gap {
        if byte == b'\n' {
            if current_line == line {
                return col;
            }
            current_line += 1;
            col = 0;
        } else {
            col += 1;
        }
    }
    
    col
}

